### ISSUE TITLE
std::scoped_lock and warning ctad-maybe-unsupported

### ISSUE THREAD

**Issue by fekir** [llvm/llvm-project#211547]
`std::scoped_lock` supports CTAD by design, yet clang complains that it might not be the case

```c++
#include <mutex>

void foo(){
  std::mutex m;
  std::scoped_lock lock( m );
}
```

Example on godbolt: https://compiler-explorer.com/z/von4463eK

Since `std::scoped_lock` comes from the standard library, I cannot add deduction guides to make clang happy, I would expect clang to "know" that `std::scoped_lock lock( m );` is not problematic (if it compiles), no matter from which standard library it comes.

**Comment by EugeneZelenko** [comment]
@fekir: See current issue markup for proper code block formatting.

**Comment by fekir** [comment]
> See current issue markup for proper code block formatting.

Thank you.
I was not notified that you edited the issue and wondered what you meant.

**Comment by AaronBallman** [comment]
I believe this is by design, but a bit unfortunate in the case of a standards feature (and maybe we want the design to change).

`-Wctad-maybe-unsupported` tells you when you're using an implicit deduction guide rather than an explicit one, because implicit ones can cause surprising behavior. libc++ uses an explicit deduction guide for scoped_lock which is why it doesn't produce a diagnostic, but libstdc++ relies on the implicit deduction guide: https://compiler-explorer.com/z/4YPcvGKrK That's not a bug with libstdc++, they're fine to do this, but if a Clang user opts into the `-Wctad-maybe-unsupported` diagnostic (off by default), then we'd expect to see a diagnostic for it.

That said, more research is needed to verify why the diagnostic was added in the first place. I have a vague recollection this had more to do with warning class authors that their were getting CTAD when they might not have intended to support it, which makes the diagnostic more about the class declaration; the use is required to keep the diagnostic practical because otherwise the diagnostic would trigger for basically every templated class. If that's accurate, it might make sense to diagnose with the class declaration as the location (and note the use which triggered CTAD), so then system header suppression would kick in for the diagnostic for scoped_lock.
